### ISSUE TITLE
Pwntools runs especially slow on ubuntu compared to other distros

### ISSUE THREAD

**Issue by caprinux** [Gallopsled/pwntools#1899]
Consider the attached zip file; [help.zip](https://github.com/Gallopsled/pwntools/files/6565898/help.zip)

On migrating from Kali to Ubuntu 20.04.2.0 LTS, I have noticed that pwntools runs significant slowly for local scripts on Ubuntu as compared to other linux distros. 

Here is the result of the same script being ran on Ubuntu 20.04.2.0 LTS  VMWare versus Kali Linux Virtualbox

Ubuntu 20.04.2.0 LTS VMware:

![image](https://user-images.githubusercontent.com/76640319/120097118-5bbc7d00-c161-11eb-952f-437d4b327e4d.png)

Kali Linux VirtualBox:

![image](https://user-images.githubusercontent.com/76640319/120097115-565f3280-c161-11eb-9ce0-3b808d889e2a.png)

Running more tests, I have also experimented with other flavors as well:

Ubuntu 18.04 VMware : 2 minutes

Ubuntu Vagrant (VBOX) : 1 minute (attributed to the tester's faster pc but still significantly slow)

Debian Buster (Clean Install) : 5 seconds

Ubuntu WSL1 : 9 seconds

Furthermore, I have tried it on both v4.7.0 (dev) and v4.5.0 (stable) and the same issue persists. 

Hence  the only common cause spotted was the use of Ubuntu Desktop, which runs pwntools rather slowly.

Unable to get to the bottom of the issue so far, please do look into this and let me know if theres anything I can do to get to the bottom of this issue.

**Comment by theKidOfArcrania** [comment]
From a first glance of the issue, maybe your issue might be as a result of the `/proc/sys/kernel/core_pattern`, which might be running some program that slow down the entire thing for each crash that happened, try changing it to be `core` instead?

**Comment by heapcrash** [comment]
Can confirm this is pretty slow on Docker / Ubuntu, even with:

```
pwntools@741fdbe01198:~/issue-1899$ cat /proc/sys/kernel/core_pattern
core
```

Working on getting a Buster environment up to verify that it's faster.

**Comment by heapcrash** [comment]
Confirmed it's much faster on Buster.

Not really sure what there's to do here for pwntools, this seems like an OS issue.  Ubuntu is based on Debian, so I'm not sure what the issue is.

**Comment by caprinux** [comment]
is there any way to check what's slowing it down? or should i file a bug report to ubuntu instead? really looking to fix this.

**Comment by heapcrash** [comment]
You could probably do some Python profiling / benchmarks to show what's being slow, that would help identify what the slowest parts are from the Pwntools side.

Syscall time profiling is probably your best bet on a systems level.  `strace` with `ff` and `-S` and `-c` are probably helpful.

**Comment by caprinux** [comment]
On Ubuntu VMware:

![image](https://user-images.githubusercontent.com/76640319/120320805-2ce70800-c315-11eb-8508-9fa11e286363.png)

[PythonProfile.txt](https://github.com/Gallopsled/pwntools/files/6576090/PythonProfile.txt)

First 20 lines: 
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    423/1    0.001    0.000  114.792  114.792 {built-in method builtins.exec}
        1    0.021    0.021  114.792  114.792 birdie.py:1(<module>)
     1904    0.041    0.000  114.083    0.060 tube.py:273(recvuntil)
     2792    0.007    0.000  113.995    0.041 tube.py:73(recv)
     2792    0.012    0.000  113.834    0.041 tube.py:173(_recv)
     2792    0.033    0.000  113.692    0.041 tube.py:130(_fillbuffer)
     2792    0.022    0.000  113.470    0.041 process.py:704(recv_raw)
     2792    0.029    0.000  113.322    0.041 process.py:748(can_recv_raw)
     2792  113.292    0.041  113.292    0.041 {built-in method select.select}
     8376    0.035    0.000    0.348    0.000 buffer.py:177(get_fill_size)
       51    0.001    0.000    0.330    0.006 __init__.py:1(<module>)
        9    0.000    0.000    0.329    0.037 __init__.py:2(<module>)
    483/3    0.002    0.000    0.317    0.106 <frozen importlib._bootstrap>:1002(_find_and_load)
    482/3    0.001    0.000    0.316    0.105 <frozen importlib._bootstrap>:967(_find_and_load_unlocked)
    467/3    0.002    0.000    0.316    0.105 <frozen importlib._bootstrap>:659(_load_unlocked)
```

On Kali VBox:

![image](https://user-images.githubusercontent.com/76640319/120321843-59e7ea80-c316-11eb-819e-ab5818dd9408.png)

[pythonprofilekali.txt](https://github.com/Gallopsled/pwntools/files/6576129/pythonprofilekali.txt)

First 20 lines:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    425/1    0.001    0.000    4.667    4.667 {built-in method builtins.exec}
        1    0.019    0.019    4.667    4.667 birdie.py:1(<module>)
     1710    0.045    0.000    3.797    0.002 tube.py:273(recvuntil)
     2447    0.006    0.000    3.712    0.002 tube.py:73(recv)
     2447    0.012    0.000    3.590    0.001 tube.py:173(_recv)
     2444    0.031    0.000    3.466    0.001 tube.py:130(_fillbuffer)
     2444    0.018    0.000    3.269    0.001 process.py:704(recv_raw)
     2444    0.021    0.000    3.161    0.001 process.py:748(can_recv_raw)
     2444    3.139    0.001    3.139    0.001 {built-in method select.select}
        9    0.000    0.000    0.511    0.057 __init__.py:2(<module>)
    483/3    0.002    0.000    0.495    0.165 <frozen importlib._bootstrap>:1002(_find_and_load)
    483/3    0.001    0.000    0.495    0.165 <frozen importlib._bootstrap>:967(_find_and_load_unlocked)
    470/3    0.002    0.000    0.495    0.165 <frozen importlib._bootstrap>:659(_load_unlocked)
    423/2    0.001    0.000    0.495    0.247 <frozen importlib._bootstrap_external>:784(exec_module)
    626/4    0.000    0.000    0.494    0.124 <frozen importlib._bootstrap>:220(_call_with_frames_removed)
```

**Comment by caprinux** [comment]
I don't see much discrepancies here except for the insanely high average time per select syscall... Hmm any ideas?

**Comment by heapcrash** [comment]
Maybe this is something Scalene can help with?
https://github.com/plasma-umass/scalene

Also, I'd advise launching your exploit script under `trace`, not attaching by PID.

**Comment by Arusekk** [comment]
Try doing `strace -y` to see what specific file descriptors it hangs on.

**Comment by heapcrash** [comment]
@Arusekk Should we try to use `epoll` instead of select, opportunistically (and obvs only on Linux)?
https://stackoverflow.com/questions/17355593/why-is-epoll-faster-than-select/17355702

**Comment by Arusekk** [comment]
You see, if we knew what the longest select()ed file descriptors are referring to, we could inspect what the awaited events are, and what kind of data it waits for. I would not make an assumption that select() is slow because of being select(). It can be well not its fault, just that the event that it expects (input data readiness) arrives late.

-- 
Wysłane z mojego urządzenia Sailfish

**Comment by caprinux** [comment]
It is difficult to tell which file descriptors it hangs on `strace -y`, here's the bunch of syscalls that seems to get repeated at each iteration of the brute force.

![image](https://user-images.githubusercontent.com/76640319/122545222-2b7d4580-d060-11eb-9c80-0e7731f6dc23.png)

However, based on what I am seeing, the syscalls slow down at the following 2 instructions: **wait4** and a constantly flickering **select**

```
wait4(191584, 0x7fff26d5bf84, WNOHANG, NULL) = 0               
select(9, [8</dev/ptmx<char 5:2>>], [], [], {tv_sec=1048575, tv_usec=999996}
```

As for scalene, it definitely seems like  system issue.

![image](https://user-images.githubusercontent.com/76640319/122546152-3dabb380-d061-11eb-81d0-7a1b9e222787.png)

My guesses are that either the `wait4` or `select` syscall is glitching out but hm

**Comment by Arusekk** [comment]
Try doing the same again, but this time use `strace -y -f -T` so that you (y) see file descriptors, (f) follow child processes, and (T) time spent inside syscalls.

The longest syscall is `select` waiting for data from the child process. So the answer is that the child process takes ages to answer your commands.

**Comment by caprinux** [comment]
[log.txt](https://github.com/Gallopsled/pwntools/files/6676425/log.txt)

Here is an excerpt of the output from `strace -y -f -T -o log.txt python3 ./birdie.py`.

Of all the syscalls, the syscalls that takes the longest is **futex** and **select**.

I've further extracted a few of the more interesting and lengthy syscalls. Please have a look.

```
198863 <... wait4 resumed>[{WIFSIGNALED(s) && WTERMSIG(s) == SIGABRT && WCOREDUMP(s)}], 0, NULL) = 199130 <0.291532>
198864 futex(0x7f18f0001220, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, {tv_sec=1624012214, tv_nsec=370879000}, FUTEX_BITSET_MATCH_ANY) = -1 ETIMEDOUT (Connection timed out) <0.102693>
198714 <... select resumed>)            = 1 (in [8], left {tv_sec=1048575, tv_usec=736698}) <0.263465>
198864 futex(0x7f18f0001220, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, {tv_sec=1624012213, tv_nsec=542602000}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
199130 +++ killed by SIGABRT (core dumped) +++

```

**Comment by Arusekk** [comment]
So there is a time-limited futex timing out? No idea what uses such calls.

**Comment by Arusekk** [comment]
Okay, so I *think* Ubuntu has some patch to do a random wait, in order to mess with timing attacks, in `__stack_chk_fail`... it would be unavoidable then, and 100% Ubuntu-specific, probably coming soon to other distros. Not sure, though.

**Comment by caprinux** [comment]
Ah so there's no way to reprogram this or find a workaround? That's rather unfortunate then hm

**Comment by heapcrash** [comment]
@Arusekk do you have any references / reading material?

**Comment by Arusekk** [comment]
@heapcrash nope, just a wild guess

**Comment by peace-maker** [comment]
Hm, I couldn't reproduce the issue on Ubuntu 20.04.3 in Virtualbox anymore with `core_pattern` set to `core`. It would run even faster with core files disabled completely. Did this just fix itself between 18.04 and 20.04?

**Comment by caprinux** [comment]
Just tested this out, it indeed works perfectly as intended with `core_pattern` set to `core` or even better, `none`.

In hindsight, this does make sense because the exploit is brute-forcing the canary of the program which causes the threads to constantly be SIGABRT'ed and hence creating a core for each round of brute-forcing and hence slowing down the program tremendously. 

This conclusion does make the issue redundant as it is an intended action caused by the system which is out of our reach. Thanks for your time, I will be closing this issue :)
